### PR TITLE
doc update: how to use streamlined auth-flow

### PR DESCRIPTION
**Pull request by abeatrix**
Close: https://github.com/sourcegraph/sourcegraph/issues/36728

Update docs with details on how the streamlined auth flow introduced in https://github.com/sourcegraph/sourcegraph/pull/35339 works, and how it can be implemented effectively by other extensions.

A copy of the diagram for the streamlined auth flow:

PS: This is the first time I've created a diagram so any feedback is greatly appreciated 😄 

```mermaid
sequenceDiagram
    Extension (Client App)->>Sourcegraph Auth UI (Browser): Open auth URL with the<br/>`requestFrom` URL param
    Sourcegraph Auth UI (Browser)->>Token Creation Callback Page: Redirect authenticated users
    Token Creation Callback Page->>Token Creation Callback Page: Verify extension
    Note right of Token Creation Callback Page: Check if the allow list has<br/>the `requestFrom` value
    Token Creation Callback Page->>Web Server: Request token for verified extension
    Web Server->>Token Creation Callback Page: Create and return token
    Token Creation Callback Page->>Sourcegraph Auth UI (Browser): Present token
    Token Creation Callback Page->>Extension (Client App): Return token with redirect URL
    Note right of Token Creation Callback Page: Open pre-authorized redirect URL<br/>for the verified extension with<br/>token added to the `code` URL params
    Extension (Client App)->>Extension (Client App): Import token
    Note right of Extension (Client App): Retrieve token from URL param<br>using a custom URL handler
``` 

CC @philipp-spiess @vdavid for jetbrains

## Test plan

N/A - Docs update

## App preview:

- [Web](https://sg-web-bee-docs-auth.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-cdqmtrddwl.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: doc/dev/how-to/index.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 - [How to add support for a language](add_support_for_a_language.md)
 - [Release Browser Extensions](releasing_browser_extensions.md)
+- [How to simplify the authentication process for an IDE extension](streamlined_auth_flow.md)
```
Comment (tjkandala) [inline doc/dev/how-to/index.md:11]:
```suggestion
- [How to implement the authentication process for an IDE extension](ide_auth_flow.md)
```
I think "simplification" only makes sense for existing IDE extensions, so we could rename the file to reflect that for when we make the neovim extension, for example.

File: doc/dev/how-to/streamlined_auth_flow.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
+# Authentication for IDE
+
+This document will go through the current auth flow for our IDE extensions, and how to implement them in other extensions.
+
+## Personal Access Token
+
+In general, users are required to provide each IDE extension with their Sourcegraph Personal Access Token to make certain requests to the Sourcegraph API.
+
+Before the token creation callback page was introduced, users must retrieve a Sourcegraph Personal Access Token from their Sourcegraph instance manually in order to proceed with the authentication process for each IDE extension.
```
Comment (tjkandala) [inline doc/dev/how-to/streamlined_auth_flow.md:9]:
```suggestion
Before the token creation callback page was introduced, users had to retrieve a Sourcegraph Personal Access Token from their Sourcegraph instance manually in order to proceed with the authentication process for each IDE extension.
```

File: doc/dev/how-to/streamlined_auth_flow.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+Steps to authenticate an account for an extension:
```
Comment (tjkandala) [inline doc/dev/how-to/streamlined_auth_flow.md:11]: Something like "default steps" or "old steps" may help to make it clear that this is the old way to do it?

File: doc/dev/how-to/streamlined_auth_flow.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+> - http://localhost:8323?code=$TOKEN
+> - The redirect URL is generated with the redirect URL listed under the verified source from the [allow list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7533ada4fd47e82d95f7258cbb88ca76f414e53d/-/blob/client/web/src/user/settings/accessTokens/UserSettingsCreateAccessTokenCallbackPage.tsx?L59-69).
+
+The token should then be processed by the extension's URL handler.
+
+### Architecture Diagram
+
+```mermaid
+sequenceDiagram
+    Extension (Client App)->>Sourcegraph Auth UI (Browser): Open auth URL with the<br/>`requestFrom` URL param
```
Comment (tjkandala) [inline doc/dev/how-to/streamlined_auth_flow.md:55]:
```suggestion
    Extension->>Sourcegraph Auth UI (Browser): Open auth URL with the<br/>`requestFrom` URL param
```
Just some thoughts on "participant" names:
- Both the extension and web app are clients, so we can just call it `Extension`
- `Token Creation Callback Page` isn't annotated with (Browser), but it's probably clear from context that it too is in the browser?

**Review by tjkandala — APPROVED** [review]
Awesome documentation and diagram!! I'll defer to the Jetbrains crew to see if it makes sense to other consumers, but it looks to me

File: doc/dev/how-to/streamlined_auth_flow.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+> EXAMPLE: Redirect URL from the Token Creation Callback Page to the extension: 
+> - http://localhost:8323?code=$TOKEN
+> - The redirect URL is generated with the redirect URL listed under the verified source from the [allow list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7533ada4fd47e82d95f7258cbb88ca76f414e53d/-/blob/client/web/src/user/settings/accessTokens/UserSettingsCreateAccessTokenCallbackPage.tsx?L59-69).
+
+The token should then be processed by the extension's URL handler.
+
+### Architecture Diagram
+
+```mermaid
```
Comment (vdavid) [inline doc/dev/how-to/streamlined_auth_flow.md:53]: +1 for using mermaid for this :)

**Review by vdavid — APPROVED** [review]
Nice stuff! Thanks for making this generic so that we won't need to change a lot of things around once we incorporate the process into JetBrains. <3

One suggestion/thought: I think I'd also add the description of the old/manual process, to cover the weird cases when a company has an older version or if this process fails for whatever reason. I know that the docs are in the same repo and it's shipped together with the new version, but the real world is quirky, so who knows, maybe a company uses two Sourcegraph versions in parallel, or a developer uses the docs on sourcegraph.com but their company uses an old version, etc.

**Pull request merged**